### PR TITLE
feat(css_parser):simple block 

### PR DESCRIPTION
**Pull request by suxin2017**
https://github.com/biomejs/biome/issues/268
## Summary

Supports simple css block implementation
support simple func
support declaration list
Support the content

! // is not supported
```
CssDeclaration =
name: CssIdentifier | CssCustomProperty
':'
value: AnyCssValue
important: CssDeclarationImportant?

AnyCssValue =
CssIdentifier
| CssString
| CssNumber
| CssDimension
| CssRatio
| CssAnyFunction
| CssCustomProperty
```

## Test Plan

cargo test -p biome_css_parser -- --test   spec_test::quick_test  --ignored

### CODE REVIEW

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 1]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
+pub(crate) fn parse_declaration_item(p: &mut CssParser) {
+    if !is_at_identifier(p) {
+        return;
+    }
+    let m = p.start();
+    // name
+    parse_regular_identifier(p).ok();
+
+    p.expect(T![:]);
+    loop {
```
Comment (denbezrukov) [inline crates/biome_css_parser/src/syntax/mod.rs:120]: Why do we need loop here?
Is it possible to ha a list here?

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 2]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
+        || p.at(CSS_STRING_LITERAL)
+        || is_at_css_dimension(p)
+        || p.at(CSS_NUMBER_LITERAL)
+        || is_at_css_custom_property(p)
+        || is_at_css_ratio(p)
+}
+
+#[inline]
+pub(crate) fn parse_any_css_value(p: &mut CssParser) -> ParsedSyntax {
+    let css_any_function = parse_css_any_function(p);
```
Comment (denbezrukov) [inline crates/biome_css_parser/src/syntax/mod.rs:149]: I'd like to propose the following structure:
```
if is_at_ration(p) {
  parse_ration(p)
} else if is_at_custom_property(p) {
  parse_custom_property(p)
}
```

File: crates/biome_css_parser/src/lexer/mod.rs [hunk 3]
```diff
@@ -944,7 +946,7 @@ impl<'src> CssLexer<'src> {
 
                 COMMENT
             }
-            _ => self.consume_unexpected_character(),
+            _ => self.consume_byte(T![/]),
```
Comment (denbezrukov) [inline crates/biome_css_parser/src/lexer/mod.rs:949]: Could you please explain why we need to consume `T![/]`?

**Review by denbezrukov — CHANGES_REQUESTED** [review]
Great start with the CSS block! I have some questions and proposals to discuss.

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 4]
```diff
@@ -141,3 +322,16 @@ pub(crate) fn parse_string(p: &mut CssParser) -> ParsedSyntax {
 
     Present(m.complete(p, CSS_STRING))
 }
+
+#[inline]
+pub(crate) fn parse_css_number(p: &mut CssParser) -> ParsedSyntax {
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs]: Yes, I'll remove this.

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 5]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
+}
+
+#[inline]
+pub(crate) fn is_at_css_ratio(p: &mut CssParser) -> bool {
+    p.at(CSS_NUMBER_LITERAL) && p.nth_at(1, T![/])
+}
+
+#[inline]
+pub(crate) fn parse_css_ratio(p: &mut CssParser) -> ParsedSyntax {
+    if is_at_css_ratio(p) {
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:258]: ok

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 6]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
+        simple_fn.complete(p, CSS_SIMPLE_FUNCTION);
+
+        return Present(m.complete(p, CSS_ANY_FUNCTION));
+    }
+    Absent
+}
+
+#[inline]
+pub(crate) fn is_at_css_dimension(p: &mut CssParser) -> bool {
+    p.at(CSS_NUMBER_LITERAL) && matches!(p.nth(1), T![%] | T![ident])
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:236]: Do you have any suggestions?

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 7]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
+    loop {
+        if !p.eat(T![;]) {
+            break;
+        }
+        parse_declaration_item(p);
+    }
+    m.complete(p, CSS_DECLARATION_LIST);
+}
+
+// CssDeclaration =
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:105]: I don't think it's necessary, it's just handy to remind me when I'm in development

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 8]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
+        if any_css_value.is_absent() {
+            break;
+        }
+    }
+    parse_declaration_important(p);
+    m.complete(p, CSS_DECLARATION);
+}
+
+#[inline]
+pub(crate) fn parse_declaration_important(p: &mut CssParser) {
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:131]: ok

File: crates/biome_css_parser/src/syntax/mod.rs (hunk 1 above)
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:120]: I think this should be done with ParseSeparatedList just like the one above, don't you think?

File: crates/biome_css_parser/src/syntax/mod.rs (hunk 2 above)
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:149]: ok

File: crates/biome_css_parser/src/lexer/mod.rs (hunk 3 above)
Comment (suxin2017) [reply crates/biome_css_parser/src/lexer/mod.rs:949]: in css ratio 1 / 3

File: crates/biome_css_parser/src/syntax/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (denbezrukov) [reply crates/biome_css_parser/src/syntax/mod.rs:226]: We can try to create a list of valid identifiers:
It seems this list is exhausted:
https://github.com/swc-project/swc/blob/d0a885d3584d3817733567e6af6e4f46e1a81964/crates/swc_css_parser/src/parser/values_and_units/mod.rs#L3614-L3647


https://github.com/swc-project/swc/blob/d0a885d3584d3817733567e6af6e4f46e1a81964/crates/swc_css_parser/src/parser/values_and_units/mod.rs#L1965-L1984

File: crates/biome_css_parser/src/syntax/mod.rs (hunk 1 above)
Comment (denbezrukov) [reply crates/biome_css_parser/src/syntax/mod.rs:117]: Do we need parse a list here? 
Our gramma doesn't have a list as value now:

```
CssDeclaration =
	name: CssDeclarationName
	':'
	value: AnyCssValue
	important: CssDeclarationImportant?
```

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 9]
```diff
@@ -79,13 +80,192 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
...
     p.expect(T!['}']);
 
     Present(m.complete(p, CSS_BLOCK))
 }
 
+pub(crate) fn parse_declaration_list(p: &mut CssParser) {
+    let m = p.start();
+
+    // first parse declaration item
+    parse_declaration_item(p);
```
[the remark this answers is not in the document]
Comment (denbezrukov) [reply crates/biome_css_parser/src/syntax/mod.rs:95]: You could find an example here:
https://github.com/biomejs/biome/blob/b71b4d0f4e983e2d8081a9ef5889191a59c7a162/crates/biome_css_parser/src/syntax/selector/mod.rs#L65-L98

File: crates/biome_css_parser/src/syntax/mod.rs (hunk 1 above)
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs:117]: I think this grammar is wrong
I will fix them tomorrow
The following is the correct grammar
`flex: 1 1 0`
` background: url(something) 100%`
It is a compent value list
https://drafts.csswg.org/css-syntax/#consume-declaration

**Comment by suxin2017** [comment]
Hello, I have finished the review of the todo, can you help me to review again @denbezrukov

File: crates/biome_css_parser/src/syntax/css_dimension.rs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+fn parse_css_unit(p: &mut CssParser) -> ParsedSyntax {
+    if !(p.at(T![ident]) && is_unit_str(p.cur_text())) {
+        return Absent;
+    }
+    parse_regular_identifier(p)
+}
+
+#[inline]
+fn is_unit_str(unit: &str) -> bool {
+    is_length_unit(unit)
```
Comment (denbezrukov) [inline crates/biome_css_parser/src/syntax/css_dimension.rs:64]: I'm wondering if we need to have "em", "rem" as keywords. 
See 
https://github.com/biomejs/biome/blob/6739a5148b41e3d81b19722fffe17d97de4434dc/xtask/codegen/src/css_kinds_src.rs#L47
https://github.com/biomejs/biome/blob/6739a5148b41e3d81b19722fffe17d97de4434dc/xtask/codegen/css.ungram#L223
https://github.com/biomejs/biome/blob/6739a5148b41e3d81b19722fffe17d97de4434dc/crates/biome_css_parser/src/lexer/mod.rs#L727-L770

We can move them to keywords in another PR.

**Review by denbezrukov — APPROVED** [review]
Thank you for your contribution! We need to address several issues, but overall, it looks good and is a valuable addition to the project.

File: xtask/codegen/css.ungram [hunk 11]
```diff
@@ -399,15 +399,17 @@ CssBlock =
...
-	name: CssIdentifier | CssCustomProperty
+	name: CssDeclarationName
 	':'
-	value: AnyCssValue
+	value: CssListOfComponentValues
 	important: CssDeclarationImportant?
 
+CssListOfComponentValues = AnyCssValue*
+
+CssDeclarationName = CssIdentifier | CssCustomProperty
```
Comment (denbezrukov) [inline xtask/codegen/css.ungram:412]: It's better to rename to `AnyCssDeclarationName`

File: crates/biome_css_parser/src/syntax/css_dimension.rs (hunk 10 above)
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/css_dimension.rs:64]: I tried but got the following error, we can fix it in the next pr.
```
-> biome\crates\biome_parser\src\token_set.rs:46:18
   |
46 |         _ => [0, 1u128 << (num - 127)],
   |                  ^^^^^^^^^^^^^^^^^^^^ attempt to shift left by `170_usize`, which would overflow
```

We may need a modification such as the following
```rust
--- a/crates/biome_parser/src/token_set.rs
+++ b/crates/biome_parser/src/token_set.rs
@@ -2,10 +2,10 @@ use biome_rowan::SyntaxKind;
 use std::marker::PhantomData;

 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub struct TokenSet<K: SyntaxKind>([u128; 2], PhantomData<K>);
+pub struct TokenSet<K: SyntaxKind>([u128; 3], PhantomData<K>);

 impl<K: SyntaxKind> TokenSet<K> {
-    pub const EMPTY: TokenSet<K> = TokenSet([0; 2], PhantomData);
+    pub const EMPTY: TokenSet<K> = TokenSet([0; 3], PhantomData);

     pub fn singleton(kind: K) -> Self {
         unsafe { TokenSet::from_raw(kind.to_raw().0) }
@@ -13,7 +13,7 @@ impl<K: SyntaxKind> TokenSet<K> {

     pub const fn union(self, other: TokenSet<K>) -> Self {
         TokenSet(
-            [self.0[0] | other.0[0], self.0[1] | other.0[1]],
+            [self.0[0] | other.0[0], self.0[1] | other.0[1],self.0[2] | other.0[2]],
             PhantomData,
         )
     }
@@ -23,7 +23,8 @@ impl<K: SyntaxKind> TokenSet<K> {
         let num = kind as usize;
         match num {
             0..=127 => self.0[0] & mask(kind)[0] != 0,
-            _ => self.0[1] & mask(kind)[1] != 0,
+            128..=255 => self.0[1] & mask(kind)[1] != 0,
+            _ => self.0[2] & mask(kind)[2] != 0,
         }
     }

@@ -39,11 +40,12 @@ impl<K: SyntaxKind> TokenSet<K> {
     }
 }

-const fn mask(kind: u16) -> [u128; 2] {
+const fn mask(kind: u16) -> [u128; 3] {
     let num = kind as usize;
     match num {
-        0..=127 => [1u128 << num, 0],
-        _ => [0, 1u128 << (num - 127)],
+        0..=127 => [1u128 << num, 0,0],
+        128..=255 => [0,1u128 << (num - 127), 0],
+        _ => [0,0, 1u128 << (num - 255)],
     }
 }

```

File: crates/biome_css_parser/src/syntax/mod.rs [hunk 12]
```diff
@@ -79,13 +84,246 @@ pub(crate) fn parse_rule_block(p: &mut CssParser) -> ParsedSyntax {
     }
     let m = p.start();
     p.expect(T!['{']);
-    let list = p.start();
-    list.complete(p, CSS_DECLARATION_LIST);
+    CssDeclarationList::default().parse_list(p);
+    // parse_declaration_list(p);
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/mod.rs]: Clearance completed

File: crates/biome_css_parser/src/syntax/css_dimension.rs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+fn is_at_css_percentage_dimension(p: &mut CssParser) -> bool {
+    p.at(CSS_NUMBER_LITERAL) && matches!(p.nth(1), T![%])
+}
+#[inline]
+fn parse_css_percentage_dimension(p: &mut CssParser) -> ParsedSyntax {
+    if !is_at_css_percentage_dimension(p) {
+        return Absent;
+    }
+    let m = p.start();
+    let _css_number = parse_regular_number(p);
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/css_dimension.rs]: done

File: crates/biome_css_parser/src/syntax/css_dimension.rs [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+use biome_css_syntax::T;
+use biome_parser::prelude::ParsedSyntax;
+use biome_parser::prelude::ParsedSyntax::{Absent, Present};
+use biome_parser::Parser;
+
+use super::parse_error::expected_unit;
+use super::{parse_regular_identifier, parse_regular_number};
+
+#[inline]
+pub(crate) fn is_at_css_dimension(p: &mut CssParser) -> bool {
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply crates/biome_css_parser/src/syntax/css_dimension.rs]: done

File: xtask/codegen/css.ungram (hunk 11 above)
Comment (suxin2017) [reply xtask/codegen/css.ungram]: done

File: crates/biome_css_parser/src/syntax/css_dimension.rs (hunk 10 above)
Comment (denbezrukov) [reply crates/biome_css_parser/src/syntax/css_dimension.rs:64]: Oh, I see.. 
Maybe we can remove all colours from keywords for now.
Anyway we can do it in another PR.

**Comment by denbezrukov** [comment]
@suxin2017 Thank you sincerely for your valuable contribution.

**Comment by suxin2017** [comment]
> @suxin2017 Thank you sincerely for your valuable contribution.

you are welcome

**Pull request merged**